### ISSUE TITLE
Make quarantining of the test class itself optional

### ISSUE THREAD

**Issue by seanf** [BinaryTweed/quarantining-test-runner#3]
Thanks for a handy library @BinaryTweedDeej. I'm using `QuarantiningRunner` to help with a migration away from TestNG.  The quarantining provides an alternative to TestNG's `IObjectFactory` for my purposes.  I use it to isolate `org.jboss.seam`.

I'm finding the automatic isolation of the actual test class to be a bit of a problem, because it leads to problems when using non-public (package private) methods from other (non-test) classes in the same package.  

> java.lang.IllegalAccessError: tried to access method org.zanata.dao.TextFlowTargetHistoryDAO.convertTimeZoneFunction(Ljava/lang/String;Lcom/google/common/base/Optional;Lorg/joda/time/DateTimeZone;)Ljava/lang/String; from class org.zanata.dao.TextFlowTargetHistoryDAOTest

If (as a workaround) I try to isolate the entire package, I get error messages like these:

> 2015-06-01 12:05:42,795 main ERROR [QuarantiningUrlClassLoader] - Could not load [org.zanata.model.ModelEntityBaseBeanInfo]
> 2015-06-01 12:05:42,796 main ERROR [QuarantiningUrlClassLoader] - Could not load [org.zanata.model.ModelEntityBaseCustomizer]

followed by a `ClassCastException` regarding `org.zanata.model.ModelEntityBase`.

`org.zanata.model.ModelEntityBase` is one of my classes, but the other classes are generated by Hibernate.  One of the other class loaders manages to load these classes, but when run they fail to access package private members because of the different class loader.

If I could avoid quarantining the test class itself, I think I could work around this.  Alternatively, if I could use exclusion patterns, I think I could isolate all of `org.zanata`, but excluding `org.zanata.model` and its Hibernate-related classes.

**Comment by seanf** [comment]
I forgot to mention: my current workaround for that class is to add `org.zanata.dao.TextFlowTargetHistoryDAO` to the quarantine.

**Comment by BinaryTweedDeej** [comment]
Hey @seanf! Really glad you're finding the tool useful.

I'd like to write a failing test to represent your use case and so understand it better.

Given a class in the 'main' classpath:

``` java
package org.zanata.model;

Imports...

public class ReferencedClass {
   String getFoo() {
      return "foo";
   }
}
```

If you call it from a quarantined test, then an `IllegalAccessError` gets thrown:

``` java
package org.zanata.model.ReferencingTest

Imports...

@RunWith(QuarantiningRunner.class)
@Quarantine({"org.jboss.seam"})
public class ReferencingTest {
   @Test
   public void packagePrivateReferenceThrowsException() {
      ReferencedClass referenced = new ReferencedClass();
      assertThat(referenced.getFoo(), is("foo));
      //Will blow up on the line above
   }
}
```

Have I understood the problem correctly?

**Comment by seanf** [comment]
@BinaryTweedDeej That's the crux of it, yes.

My latest workaround is a modified copy of `QuarantiningRunner` which doesn't quarantine the test class (just using `quarantinedPatterns` instead of `allPatterns`), and that's working for all my tests.  My other workaround didn't scale when it came to all the other tests.

**Comment by BinaryTweedDeej** [comment]
Hey @seanf,

Just wanted to let you know I have been looking at this, but ran out of time this evening. Will hopefully get something sorted tomorrow.

**Comment by BinaryTweedDeej** [comment]
Hi @seanf,

I've written an implementation and accompanying tests on [a branch](https://github.com/BinaryTweed/quarantining-test-runner/tree/issue-3). One of the [tests is currently failing](https://travis-ci.org/BinaryTweed/quarantining-test-runner/builds/65201023), as classes specified in the pattern aren't being loaded by `QuarantiningUrlClassLoader`.

I'm a bit befuddled as to how your classes are getting loaded. If the test class (ie the one annotated `@Test`) isn't loaded with `QuarantiningUrlClassLoader`, then none of the classes _it_ loads will be loaded by it either, and so they can't be considered for quarantining. Perhaps I've missed something about your implementation?

Could you turn up logging and paste in here what you see from the `QuarantiningUrlClassLoader` please?

**Comment by seanf** [comment]
@BinaryTweedDeej Sorry, but I'm not sure what you're asking.  Which tests should I run, and which version of `QuarantiningRunner`?

**Comment by BinaryTweedDeej** [comment]
Apologies for not being clear.

Could you run your tests (or at least one of them) with your version of the code please? I'm hoping to see something in the logs that shows that the quarantined classes ("org.jboss.seam" in your case) are _actually_ getting quarantined. At the moment I don't understand how they are, as if the test class itself is not quarantined, then none of the classes it loads should be either.

If you could also quickly look at [my test](https://github.com/BinaryTweed/quarantining-test-runner/blob/issue-3/src/test/java/com/binarytweed/test/QuarantinedRunnerExcludeTestAndRunnerIntegrationTest.java#L22-L27) to see if it's representative of what you're trying to do, that'd be really helpful.

Ta!

**Comment by seanf** [comment]
Thanks.  

It's entirely possible I've made a complete mess of it, and it really is doing nothing.  The problems which led me to isolate `org.jboss.seam`  only happen in rare circumstances.  And the little unit test I wrote to simulate the problem is not really representative...

Let me see what I can find.

**Comment by BinaryTweedDeej** [comment]
Aww, it'd be a shame if it turns out to be doing nothing all along!

Another solution might be to quarantine not just the test class, but its whole package - that way all the other classes in its package would be loaded by the same `ClassLoader`, and you wouldn't get the `IllegalAccessError`.

**Comment by seanf** [comment]
You're quite right.  My `SeamQuarantineTest` was actually explicitly quarantining its own class, thus invalidating the whole thing.  I moved the Seam references out into another class, added it to the `@Quarantine`, and in the debugger it's all just `sun.misc.Launcher$AppClassLoader`, with a broken test.

**Comment by seanf** [comment]
I tried quarantining `org.zanata`, but that's when I ran into the problem with Hibernate's generated classes...

**Comment by seanf** [comment]
To avoid `IllegalAccessError` within the same package, it probably would be most correct to quarantine entire packages in all cases (including packages for test class and the delegate runner), but that still leaves the Hibernate problem.

**Comment by BinaryTweedDeej** [comment]
Hmm - I wonder what's going on with those Hibernate classes? Do they get generated (like, source/class files stuck on the filesystem), or is it dynamic in the sense that classes only exist whilst the JVM is running?

**Comment by seanf** [comment]
The latter.

**Comment by seanf** [comment]
Although I think it can be done the other way too (compile time enhancement), as an option.

**Comment by BinaryTweedDeej** [comment]
I'm guessing then that those classes exist somewhere that the `QuarantiningUrlClassLoader` can't see them. It will try loading a class itself, and if it can't, pass the buck up the chain (to the `ClassLoader` that loaded it, and so on). If the Hibernate classes are getting loaded by a `ClassLoader` that's not part of that chain (ie specified later) then that'd explain why we can't find them.

Any ideas if Hibernate is specifying its own `ClassLoader` at some point?

**Comment by seanf** [comment]
I think it must be.  Perhaps the current thread's context class loader comes into this.

**Comment by seanf** [comment]
@BinaryTweedDeej got to go, but I'll see what I can find out about Hibernate's class loader tomorrow.

**Comment by BinaryTweedDeej** [comment]
This may be of interest: https://forum.hibernate.org/viewtopic.php?f=1&t=993293

**Comment by seanf** [comment]
> This may be of interest: https://forum.hibernate.org/viewtopic.php?f=1&t=993293 

Thanks, I saw that, but I'm not sure how to make a ClassLoader which sees everything.  I tried putting the test's classloader (ie `QuarantiningUrlClassLoader`) into the thread context, in the hope that Hibernate would use it, but it seems to make no difference.

I misinterpreted the ERROR logs from `QuarantiningUrlClassLoader` above: some code tries to load the BeanInfo classes if available, but it's okay if they don't exist.  (See https://docs.oracle.com/javase/tutorial/javabeans/advanced/customization.html)  I don't think the `BeanInfo`/`Customizer` classes are being generated after all, just requested.  (I think those ERROR logs should really be DEBUG.  I think most classloaders just throw the `ClassNotFoundException`, they don't talk about it in the log.)

The `ClassCastException` was something else.  The exact error message changes if I quarantine more packacegs, but ultimately, some of Hibernate's classes are being loaded by the system classloader, and some by the QuarantiningUrlClassLoader, and I just can't work out why. 

Given that the title of this bug is totally wrong, and I can't spend any more time fighting with Hibernate, I'm closing this bug.  I have a less elegant (non-quarantine) workaround which will have to do.  But I appreciate all your help, @BinaryTweedDeej.
